timemachine/fe/plots.py

```diff
@@ -0,0 +1,197 @@
+    ticks = [0, 25, 50, 75, 100]
+    labels = [f"{t}%" for t in ticks]
+    ax.set_yticks(ticks)
+    ax.set_yticklabels(labels)
+
+    # labels and legends
+    ax.set_xlabel(r"$\lambda_i$")
+    ax.set_ylabel(r"pair BAR overlap ($\lambda_i$, $\lambda_{i+1}$)")
+    ax.legend()
+
+
+def make_overlap_summary_figure(components, lambdas, overlaps_by_lambda, overlaps_by_lambda_by_component):
+    _, (ax_top, ax_btm) = plt.subplots(2, 1, figsize=(7, 9))
+    plot_overlap_summary(ax_top, ["Overall"], lambdas, [overlaps_by_lambda])
+    plot_overlap_summary(ax_btm, components, lambdas, overlaps_by_lambda_by_component)
+    buffer = io.BytesIO()
+    plt.savefig(buffer, format="png")
+    buffer.seek(0)
+    return buffer.read()
+
+
+def make_overlap_detail_figure(components, u_kln_by_component_by_lambda, temperature, prefix):
+    """Make (n_lambdas - 1) x (n_components + 1) overlap plots, and return related diagnostics
+
+    Parameters
+    ----------
+    components : list of strings
+        component names
+    u_kln_by_component_by_lambda: [L,P,2,2,T] array
+    temperature: float
+        kelvin
+    prefix: string
+        beginning of plot titles
+
+    Returns
+    -------
+    overlap_detail_png: bytes
+    overlaps_by_lambda: list of floats
+    dG_errs_by_lambda_by_component: array
+    overlaps_by_lambda_by_component: array
+
+    Notes
+    -----
+    * May be a useful source of hints, but can be misleading in simple cases
+        https://github.com/proteneer/timemachine/issues/923
+    """
+    num_energy_components = len(components)
+    assert num_energy_components == u_kln_by_component_by_lambda[0].shape[0]
+
+    num_lambdas = len(u_kln_by_component_by_lambda)
+
+    num_rows = num_lambdas - 1  # one per adjacent pair
+    num_cols = num_energy_components + 1  # one per component + one for overall energy
+
+    figure, all_axes = plt.subplots(num_rows, num_cols, figsize=(num_cols * 5, num_rows * 3))
+    if num_rows == 1:
+        all_axes = [all_axes]
+
+    # free energy analysis + plots (not relying on energy decomposition)
+    kBT = BOLTZ * temperature
+    beta = 1 / kBT
+
+    all_dGs = []
+    all_errs = []
+    for lamb_idx, u_kln_by_component in enumerate(u_kln_by_component_by_lambda):
+        u_kln = u_kln_by_component.sum(0)
+
+        w_fwd = u_kln[1, 0] - u_kln[0, 0]
+        w_rev = u_kln[0, 1] - u_kln[1, 1]
+
+        # BAR
+        df, df_err = bar_with_bootstrapped_uncertainty(w_fwd, w_rev)  # reduced units
+        dG, dG_err = df / beta, df_err / beta  # kJ/mol
+        # TODO: possibly accept list of dG, dG_err as input, instead of computing on the fly?
+
+        all_dGs.append(dG)
+        all_errs.append(dG_err)
+
+        # add to plot
+        plot_axis = all_axes[lamb_idx - 1][num_energy_components]
+        plot_title = f"{prefix}_{lamb_idx - 1}_to_{lamb_idx}"
+        plot_BAR(df, df_err, w_fwd, w_rev, plot_title, plot_axis)
+
+        # TODO: should this print be moved back into simulation loop?
```

Comment:
I think it's useful to have in the loop

Reply:
I don't know how to do this in the current PR, while also separating the scheduling/execution of simulations from the execution/plotting of the pair BAR analysis.

I imagine there will probably be a couple more kinds of on-the-fly debugging prints in place of this one in the near future:

* "simulation at lambda=0.25 completed in X minutes (X ns/day)"
* "overlaps between adjacent pairs of the completed windows lambda=[0.0, 0.25, 1.0] are overlap=[0.1, 1e-100]... the pair with the lowest overlap is (0.25, 1.0), so launching a simulation at lambda=0.625"
* "free energy difference for 0.0 -> 0.25 is X +/- X kBT"

(modulo some important details)

Reply:
Ok. Let's defer this to later.